examples/src/chains/conversational_qa_built_in_memory.ts

```diff
@@ -1,44 +1,160 @@
+import { Document } from "langchain/document";
 import { ChatOpenAI } from "langchain/chat_models/openai";
-import { ConversationalRetrievalQAChain } from "langchain/chains";
+import { LLMChain } from "langchain/chains";
 import { HNSWLib } from "langchain/vectorstores/hnswlib";
 import { OpenAIEmbeddings } from "langchain/embeddings/openai";
 import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
 import { BufferMemory } from "langchain/memory";
-
 import * as fs from "fs";
+import { PromptTemplate } from "langchain/prompts";
+import { RunnableSequence } from "langchain/schema/runnable";
+import { BaseMessage } from "langchain/schema";
 
 export const run = async () => {
   const text = fs.readFileSync("state_of_the_union.txt", "utf8");
+
   const textSplitter = new RecursiveCharacterTextSplitter({ chunkSize: 1000 });
   const docs = await textSplitter.createDocuments([text]);
+
   const vectorStore = await HNSWLib.fromDocuments(docs, new OpenAIEmbeddings());
+  const retriever = vectorStore.asRetriever();
+
+  const memory = new BufferMemory({
+    memoryKey: "chatHistory",
+    inputKey: "question", // The key for the input to the chain
+    outputKey: "text", // The key for the final conversational output of the chain
+    returnMessages: true, // If using with a chat model (e.g. gpt-3.5 or gpt-4)
+  });
+
+  const serializeDocs = (docs: Array<Document>): string =>
+    docs.map((doc) => doc.pageContent).join("\n");
+
+  const serializeChatHistory = (chatHistory: Array<BaseMessage>): string =>
+    chatHistory
+      .map((chatMessage) => {
+        if (chatMessage._getType() === "human") {
+          return `Human: ${chatMessage.content}`;
+        } else if (chatMessage._getType() === "ai") {
+          return `Assistant: ${chatMessage.content}`;
+        } else {
+          return `${chatMessage.content}`;
+        }
+      })
+      .join("\n");
+
+  /**
+   * Create two prompt templates, one for answering questions, and one for
+   * generating questions.
+   */
+  const questionPrompt = PromptTemplate.fromTemplate(
+    `Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.\n\nCONTEXT: {context}\n\nCHAT HISTORY: {chatHistory}\n\nQUESTION: {question}\n\nHelpful Answer:`
```

Comment:
Would use actual newlines instead of \n for readability here - you can put declare a const QUESTION_PROMPT_TEMPLATE; variable above and remove const run () =>  as it's no longer necessary